### PR TITLE
test: add a pytest suite for the src/ pipeline functions (#6)

### LINKED ISSUE #6: Unit-Tests für zentrale Funktionen ergänzen
**Priorität:** P1 (Struktur/Professionalität)
**Betrifft:** neu: `tests/`, sowie `src/*.py` (siehe Issue 03 (#4) — Umbenennung von `03_src/`)

## Kontext

Aktuell gibt es im gesamten Repo keinerlei Tests — kein `tests/`-Verzeichnis, kein
`pytest`, kein `unittest`. Für ein Portfolio-Projekt im Bank-/Data-Science-Kontext ist das
ein auffälliger Mangel: Testabdeckung (auch nur für die Kernlogik) signalisiert
Software-Engineering-Reife und Sorgfalt, die über reines Notebook-Prototyping hinausgeht.

**Voraussetzung:** Diese Issue baut auf Issue 03 (Modularisierung, #4) auf — Tests ergeben
erst Sinn, sobald Logik als importierbare Funktionen in `src/` vorliegt.

## Ziel

Eine Basis-Testsuite (`pytest`) für die wichtigsten, deterministisch testbaren Funktionen
aus `src/` — nicht für das gesamte Notebook, sondern gezielt für die Bausteine, bei
denen ein Test echten Mehrwert hat (Datenverarbeitung, Feature Engineering, Metrikberechnung).

## Umsetzungsschritte

1. `tests/`-Verzeichnis im Repo-Root anlegen, mit `test_data_processing.py`,
   `test_features.py`, `test_evaluate.py` (passend zur Modulstruktur aus Issue 03, #4).
2. `pytest` und ggf. `pytest-cov` zu `requirements.txt` hinzufügen (siehe Issue 06, #7).
3. Für `rm_nas(df, threshold=0.4)` (bzw. dessen Nachfolger in `data_processing.py`)
   mindestens folgende Fälle testen:
   - Spalten mit Missing-Anteil über dem Threshold werden korrekt entfernt.
   - Spalten mit Missing-Anteil unter dem Threshold bleiben erhalten.
   - Edge Case: leerer DataFrame, DataFrame ohne fehlende Werte.
4. Für die Leakage-Spalten-Filterung (aus Issue 01, #2) testen, dass alle bekannten
   Post-Outcome-Spalten zuverlässig entfernt werden, auch wenn nicht alle im
   Input-DataFrame vorhanden sind (kein Fehler bei fehlenden Spalten).
5. Für Feature-Engineering-Funktionen: mit einem kleinen synthetischen Beispiel-DataFrame
   (nicht dem echten, großen Datensatz) prüfen, dass Encodings/abgeleitete Features wie
   erwartet berechnet werden. Für DataFrame-/Series-Vergleiche
   `pandas.testing.assert_frame_equal` bzw. `assert_series_equal` verwenden statt
   handgeschriebener Vergleichsschleifen.
6. Für Evaluation-Funktionen (sofern in `evaluate.py` eigene Berechnungen stattfinden,
   z. B. KS-Statistik aus Issue 07 (#8)): mit bekannten Beispielwerten (y_true, y_pred_proba)
   die korrekte Metrikberechnung verifizieren.
7. `pytest` lokal ausführen und sicherstellen, dass alle Tests grün sind.
8. Kurzen Abschnitt in der README ergänzen: wie man die Tests ausführt
   (`pytest tests/`).

## Betroffene Dateien

- Neu: `tests/test_data_processing.py`, `tests/test_features.py`,
  `tests/test_evaluate.py`
- `requirements.txt`
- `README.md` (Testausführung dokumentieren)

## Akzeptanzkriterien / Definition of Done

- [ ] `tests/`-Verzeichnis mit mindestens 3 Testdateien existiert.
- [ ] Mindestens die `rm_nas`/NA-Filter-Funktion und die Leakage-Spalten-Filterung sind
      mit mehreren Fällen (inkl. Edge Cases) abgedeckt.
- [ ] `pytest` läuft lokal ohne Fehler durch.
- [ ] README beschreibt, wie Tests ausgeführt werden.

### PR DESCRIPTION
**Pull request by FelixSchramm**
## Summary

The repository had no tests at all. This adds a `tests/` suite for the deterministic building blocks in `src/` — the cleaning steps, the feature engineering and the metric computation — plus the `pytest.ini` that makes `pytest` work from the repository root and a README step describing how to run it. 18 tests, all green.

## Type of change
- [ ] Bugfix
- [ ] New feature / analysis
- [ ] Refactoring (keine Verhaltensänderung)
- [x] Documentation — README gets the "run the tests" step
- [ ] Model/metrics change

No production code is touched: `src/` and the notebook are unchanged.

## Related issue

Issue #6. `Closes #6` does not fire here because this PR targets the integration branch, not the default branch — the reviewer closes the issue manually after the merge.

## Changes

**`tests/test_data_processing.py`** (7 tests)
- `rm_nas`: a column 60% missing is dropped while one 20% missing stays; a column at exactly 40% stays (the threshold is exclusive — `per > threshold` — which is the kind of boundary that gets flipped by accident); complete data comes back unchanged; both empty shapes (no rows and no columns, columns but no rows) are handled rather than raising.
- `drop_leakage_columns`: post-outcome fields, `hardship_*`/`settlement_*` prefixes and irrelevant identifier columns all go in one call, while `loan_amnt`, `target` and — the case worth pinning — `fico_range_low` survive. That last one is one character away from the dropped `last_fico_range_low` and is the origination score the model is allowed to see, so a sloppier substring match in `drop_leakage_columns` would silently cost a feature. A second test covers the missing-column case the issue names explicitly: by the time this function runs, most deny-listed columns are already gone via `rm_nas`, so `errors="ignore"` has to hold.
- `create_target`: running and late loans are excluded rather than labelled as non-defaults, and `Charged Off` maps to 1.

**`tests/test_features.py`** (6 tests)
- `convert_emp_length` parametrized over `< 1 year`, `1 year`, `3 years`, `10+ years` and `NaN`.
- `engineer_features` against an explicit expected frame with `assert_frame_equal` (as step 5 of the issue asks): `term` becomes `36`/`60`, `emp_length` becomes years, and the one categorical column becomes a single `home_ownership_RENT` dummy, `OWN` being the reference category under `drop_first=True`.

**`tests/test_evaluate.py`** (5 tests)
- A small `FixedProbaModel` stands in for the fitted classifier. `evaluate_model` only calls `predict_proba` and `predict`, so this keeps the expected metrics hand-derivable instead of reading them off a model trained on synthetic data — which is what step 6 of the issue asks for ("mit bekannten Beispielwerten die korrekte Metrikberechnung verifizieren").
- Perfect separation gives AUC, Gini and KS of exactly 1.
- The overlapping case is worked out in the docstring: scores 0.1/0.6 for the repaid loans and 0.4/0.9 for the defaulted ones rank three of the four pairs correctly, so AUC is 0.75, Gini 0.5 and KS 0.5.
- `y_pred_proba` is handed back unchanged (the notebook plots and interprets it).
- One regression test pins what the PR #19 review had to verify by hand: `y_test` arrives from `train_test_split` carrying the original index, and the KS split masks a numpy array with a pandas Series, which uses values rather than labels. With an index of `[7, 42, 13, 99]` the metrics must be unchanged.

**`pytest.ini`** — three lines: `pythonpath = .` so the tests can `from src... import`, and `testpaths = tests`. The alternative was an empty root `conftest.py`, which achieves the same by side effect; the explicit config seemed better to read, and #13 can build on it in CI.

**`README.md`** — a step 4 in "How to run" (`pytest`, with a sentence on what `pytest.ini` does), plus `pytest.ini` and `tests/` in the repository layout block.

## Model/Data-Leakage-Check
- [x] Beeinflusst dieser PR Trainingsdaten, Features oder das Modell? **No.** Test-only; no file under `src/` or `01_notebooks/` is modified.
- [x] Falls ja: alte vs. neue Metriken — not applicable.
- [x] Keine neuen Features verwendet, die erst nach Kreditvergabe/Ausfall bekannt sind — the opposite: `test_drop_leakage_columns_removes_post_outcome_and_irrelevant_columns` is a guard against exactly that regression.
- [x] Ein auffällig hoher/niedriger Metrikwert wurde erklärt — the AUC of 1.0 in the perfect-separation test is a constructed input, and the docstring says so.

## How to test / Verification

```
$ pytest
18 passed, 1 warning in 1.53s
```

Run against the pinned versions from #7 — `pandas 3.0.5`, `numpy 2.4.6`, `scikit-learn 1.9.0`, `scipy 1.17.1`, `pytest 9.1.1` — installed fresh in this session, which incidentally re-confirms those pins install cleanly. `black --check` and `ruff check` pass on `tests/` and `src/`.

The expected values in `test_evaluate.py` were derived by hand (pair counting for AUC, ECDF gaps for KS) before the run, not copied from the output.

### One thing the run surfaced, deliberately not fixed here

`pytest` reports a single warning, and it comes from existing code rather than from the tests: `src/features.py:35` calls `select_dtypes(include=["object"])`, and pandas 3 warns that `object` still picks up the new `str` dtype for backward compatibility and will stop doing so. The behaviour is correct today and the fix (`include=["str", "object"]`) is a change to `src/`, which is outside this issue — issue #12 explicitly owns "Warnings gezielt behandeln". Flagged for the reviewer and recorded in `HANDOVER.md`.

### Scope decisions

- **`pytest-cov` left out.** The issue says "ggf."; nothing in the repo consumes a coverage report, and #13 has not decided whether CI publishes one. Adding it now would be a dependency on a guess — the same reasoning that kept `optbinning` out in #7.
- **No tests for `train.py` or `load_raw_data`.** `split_and_scale` and `train_model` are thin wrappers around scikit-learn, and the issue asks for the functions "bei denen ein Test echten Mehrwert hat". Testing that `train_test_split` splits, or that `read_csv` reads, tests scikit-learn and pandas rather than this repository.

## Checklist
- [x] Notebook läuft sauber von oben nach unten durch — unchanged by this PR; no notebook cell is touched
- [x] Tests laufen durch (`pytest tests/`) — 18 passed
- [x] README ggf. aktualisiert und weiterhin konsistent mit dem tatsächlichen Repo-Inhalt — new test step, and `tests/` plus `pytest.ini` added to the layout block
- [x] Keine Secrets/Zugangsdaten committet

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013KmnFqcETbyCyKAN6WDmHj

---
_Generated by [Claude Code](https://claude.ai/code/session_013KmnFqcETbyCyKAN6WDmHj)_

### CODE REVIEW

File: tests/test_evaluate.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        """:param y_pred_proba: Default probability per test row."""
+        self.y_pred_proba = np.asarray(y_pred_proba)
+
+    def predict_proba(self, x_test: pd.DataFrame) -> np.ndarray:
+        """:return: The two-column probability matrix sklearn models return."""
+        return np.column_stack([1 - self.y_pred_proba, self.y_pred_proba])
+
+    def predict(self, x_test: pd.DataFrame) -> np.ndarray:
+        """:return: The 0.5-threshold labels sklearn classifiers predict."""
+        return (self.y_pred_proba >= 0.5).astype(int)
```
Comment (FelixSchramm) [inline tests/test_evaluate.py:31]: Minor, docstring format. `CLAUDE_CODING_RULES.md` asks for `:param name:` alongside `:return:`; both stub methods document the return but not `x_test`. Worth one line each, because `x_test` being *ignored* is the whole point of the stub — the probabilities are fixed at construction, so the frame only has to exist and have the right length. Fixing this myself.

Not a finding, but recorded because I considered it: the obvious "use a package instead of hand-writing it" alternative here is `unittest.mock.Mock` with `predict_proba.return_value` / `predict.return_value`. It would be marginally shorter, but the return values still have to be built from the same `np.column_stack` expression, and the explicit class is what documents the contract `evaluate_model` actually depends on (a two-column matrix whose second column is P(default)). Keeping the stub is the right call.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_data_processing.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    assert list(rm_nas(df).columns) == ["dense"]
+
+
+def test_rm_nas_keeps_a_column_exactly_at_the_threshold():
+    """The threshold is exclusive: 40% missing is still acceptable."""
+    df = pd.DataFrame({"edge": [1.0, 2.0, 3.0, None, None]})
+    assert list(rm_nas(df).columns) == ["edge"]
+
+
+def test_rm_nas_leaves_complete_data_untouched():
```
Comment (FelixSchramm) [inline tests/test_data_processing.py:27]: Minor: this is one of only two tests in the PR without a docstring (the other is `test_convert_emp_length`), while the other eight all have one. `CLAUDE_CODING_RULES.md` says every function needs a docstring, and the inconsistency inside a single PR is the part worth closing. Fixing this myself — one line, no restating of the test name.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_features.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    ("value", "expected"),
+    [
+        ("< 1 year", 0),
+        ("1 year", 1),
+        ("3 years", 3),
+        ("10+ years", 10),
+        (np.nan, 0),
+    ],
+)
+def test_convert_emp_length(value, expected):
```
Comment (FelixSchramm) [inline tests/test_features.py:21]: Same minor docstring point as in `test_data_processing.py`. Here it actually earns its line: the parametrisation encodes two decisions that are not obvious from the function name — `< 1 year` and a missing value both collapse to `0`, and `10+ years` saturates at `10` rather than becoming `11`. Fixing this myself.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by FelixSchramm — COMMENTED** [review]
## Review — PR #23 (issue #6)

**No structural findings.** 224 lines, five files, nothing under `src/` or in the notebook touched. Three minor docstring points, all fixed by me on the branch; everything else is endorsed.

### 1. Is the code as simple and as short as possible?

Yes. Nothing here is speculative. Two things I specifically checked for and did not find:

- `_evaluate` is a three-line helper used by three of the four evaluation tests, and the fourth (`test_metrics_are_positional_for_a_non_zero_based_index`) deliberately builds its inputs inline instead of growing `_evaluate` an optional `index=` parameter. That is the correct call under "no configurability that wasn't requested" — a keyword argument with exactly one caller would have been the easy mistake here.
- `pytest.ini` is four lines and no more. The PR body names the alternative (an empty root `conftest.py`) and picks the explicit config; both work, and since #7 declined `pyproject.toml` there is no existing config file to fold this into.

### 2. Does hand-written code reimplement what a package provides?

No. `pandas.testing.assert_frame_equal` is used where step 5 of the issue asks for it, `pytest.mark.parametrize` replaces what would otherwise be hand-rolled comparison loops, and the metric expectations are asserted against `evaluate_model`'s output rather than recomputed in the test. The one hand-written object is the `FixedProbaModel` stub; I considered `unittest.mock.Mock` and concluded the stub is better here — reasoning in the inline comment.

### 3. Are the coding rules followed?

Yes, apart from the docstring points below. `black --check` and `ruff check` pass on `src/` and `tests/`; naming is snake_case; docstrings are reST.

### Verification, independent of the worker's run

Installed the pins from #7 fresh (`pandas 3.0.5`, `numpy 2.4.6`, `scikit-learn 1.9.0`, `scipy 1.17.1`, `pytest 9.1.1` — a third confirmation that those pins resolve) and got `18 passed, 1 warning`, matching the PR body exactly.

A green suite only proves the tests run, not that they would catch a regression, so I mutated `src/` one change at a time and confirmed each mutation is caught by the test that claims to guard it:

| Mutation | Caught by |
| --- | --- |
| `rm_nas`: `per > threshold` → `per >= threshold` | `test_rm_nas_keeps_a_column_exactly_at_the_threshold` |
| `drop_leakage_columns`: prefix match → substring match | `test_drop_leakage_columns_removes_post_outcome_and_irrelevant_columns` (via `fico_range_low`) |
| `evaluate_model`: `y_pred_proba` becomes a `Series` with a default 0..n index | `test_metrics_are_positional_for_a_non_zero_based_index` (raises `IndexingError`) |
| `create_target`: `Charged Off` → `Fully Paid` as the positive class | `test_create_target_keeps_completed_loans_and_encodes_the_outcome` |
| `engineer_features`: `drop_first=True` → `False` | `test_engineer_features_makes_every_column_numeric` |
| `convert_emp_length`: `10+ years` → `11` | both `test_convert_emp_length[10+ years-10]` and the frame comparison |

All six fail exactly where intended and nowhere else, and the tree was verified clean against `HEAD` afterwards. The `fico_range_low` case is the one I would call load-bearing: it is the guard that makes the #2 leakage fix hard to undo by accident.

I also re-derived the two hand-computed values rather than trusting the docstrings. For `y_true = [0, 0, 1, 1]`, scores `[0.1, 0.6, 0.4, 0.9]`: three of the four default/non-default pairs rank correctly, so AUC is 3/4 and Gini `2 * 0.75 - 1 = 0.5`; the ECDF gap is 0.5 on `[0.1, 0.4)` and again on `[0.6, 0.9)`, so KS is 0.5. Both match.

Repo hygiene around the new files is fine: `.gitignore` already covers `__pycache__/` and `*.pyc`, and pytest writes its own `.gitignore` into `.pytest_cache/`, so a first test run leaves the tree clean. I confirmed that.

### Scope decisions — all three endorsed

- **`pytest-cov` left out.** The issue says "ggf.", nothing consumes a coverage report, and #13 has not decided whether CI publishes one. Same reasoning that kept `optbinning` out of #7: a dependency added on a guess.
- **No tests for `train.py` / `load_raw_data`.** These are thin wrappers around `train_test_split`, `StandardScaler` and `read_csv`. Testing them tests scikit-learn and pandas, which is the reinvention this review is supposed to catch, not prevent.
- **The pandas warning deferred to #12.** Correct. The warning comes from `src/features.py:35`, the fix is a one-argument change to `src/`, and this is a test-only PR. #12 explicitly owns "Warnings gezielt behandeln". I reproduced the warning and it is exactly the `select_dtypes(include=["object"])` / `str` dtype deprecation described.

### One gap, deliberately not implemented here

`evaluate_model` returns five keys and four of them are asserted on; `report` is not. Almost all of that is sklearn's `classification_report` and rightly untested — but the ordering of `TARGET_NAMES` in `src/evaluate.py` is repo-owned, and if those two labels were ever swapped, every classification report in the notebook would silently attribute the default-class precision and recall to "Fully Paid (0)" with nothing failing.

Adding a test is not one of the fix categories a reviewer applies directly, and the issue does not ask for it, so I have left it alone and recorded it in `HANDOVER.md` instead — a natural pickup for #13 when CI decides what it runs, or for #12.

### Acceptance criteria

- [x] `tests/` with at least three test files
- [x] `rm_nas` and the leakage filter covered with multiple cases including edge cases (empty frame in both shapes, exact threshold, absent columns)
- [x] `pytest` runs green — 18 passed, verified independently
- [x] README documents how to run the tests

Step 2 of the issue (add `pytest` to `requirements.txt`) was already satisfied by #7, which pinned `pytest==9.1.1`.

Submitted as a comment rather than an approval because the PR is authored by the same account this session acts as, and GitHub does not allow self-approval. Merging.

---
_Generated by [Claude Code](https://claude.ai/code)_
